From af537c5c6be2b16be34f63108b57672f65646848 Mon Sep 17 00:00:00 2001
From: Eugene Zemtsov <eugene@chromium.org>
Date: Mon, 13 Jan 2025 18:12:58 -0800
Subject: [PATCH] webcodecs: Fix signed overflow in Av1VideoEncoder

Bug: 389333853
Change-Id: I90e13351041b8dd06ac754905c7d19fe6586eabc
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6171812
Reviewed-by: Thomas Guilbert <tguilbert@chromium.org>
Commit-Queue: Eugene Zemtsov <eugene@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1405855}
---

--- a/media/video/av1_video_encoder.cc
+++ b/media/video/av1_video_encoder.cc
@@ -530,7 +530,7 @@
   }
 
   bool key_frame = encode_options.key_frame;
-  auto duration_us = GetFrameDuration(*frame).InMicroseconds();
+  int64_t duration_us = GetFrameDuration(*frame).InMicroseconds();
   last_frame_timestamp_ = frame->timestamp();
   if (last_frame_color_space_ != frame->ColorSpace()) {
     last_frame_color_space_ = frame->ColorSpace();
@@ -595,14 +595,19 @@
   auto error =
       aom_codec_encode(codec_.get(), image, artificial_timestamp_, duration_us,
                        key_frame ? AOM_EFLAG_FORCE_KF : 0);
-  artificial_timestamp_ += duration_us;
-
   if (error != AOM_CODEC_OK) {
     auto msg = LogAomErrorMessage(codec_.get(), "AOM encoding error", error);
     std::move(done_cb).Run(
         EncoderStatus(EncoderStatus::Codes::kEncoderFailedEncode, msg));
     return;
   }
+  if (!base::CheckAdd(artificial_timestamp_, duration_us)
+           .AssignIfValid(&artificial_timestamp_)) {
+    std::move(done_cb).Run(EncoderStatus(
+        EncoderStatus::Codes::kEncoderFailedEncode,
+        "Timestamp overflow. Most likely, frame's duration is too large."));
+    return;
+  }
   auto output = GetEncoderOutput(std::move(temporal_id_status).value(),
                                  frame->timestamp(), frame->ColorSpace());
   if (svc_params_.number_temporal_layers > 1) {
